lib/services/saml.go

```diff
@@ -0,0 +1,191 @@
+	}
+
+	if sc.GetEntityDescriptor() != "" {
+		metadata := &types.EntityDescriptor{}
+		if err := xml.Unmarshal([]byte(sc.GetEntityDescriptor()), metadata); err != nil {
+			return nil, trace.Wrap(err, "failed to parse entity_descriptor")
+		}
+
+		for _, kd := range metadata.IDPSSODescriptor.KeyDescriptors {
+			for _, samlCert := range kd.KeyInfo.X509Data.X509Certificates {
+				certData, err := base64.StdEncoding.DecodeString(strings.TrimSpace(samlCert.Data))
+				if err != nil {
+					return nil, trace.Wrap(err)
+				}
+				cert, err := x509.ParseCertificate(certData)
+				if err != nil {
+					return nil, trace.Wrap(err, "failed to parse certificate in metadata")
+				}
+				certStore.Roots = append(certStore.Roots, cert)
+			}
+		}
+		sc.SetIssuer(metadata.EntityID)
+		if len(metadata.IDPSSODescriptor.SingleSignOnServices) > 0 {
+			sc.SetSSO(metadata.IDPSSODescriptor.SingleSignOnServices[0].Location)
+		}
+	}
+
+	if sc.GetIssuer() == "" {
+		return nil, trace.BadParameter("no issuer or entityID set, either set issuer as a parameter or via entity_descriptor spec")
+	}
+	if sc.GetSSO() == "" {
+		return nil, trace.BadParameter("no SSO set either explicitly or via entity_descriptor spec")
+	}
+	if sc.GetCert() != "" {
+		cert, err := tlsca.ParseCertificatePEM([]byte(sc.GetCert()))
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		certStore.Roots = append(certStore.Roots, cert)
+	}
+	if len(certStore.Roots) == 0 {
+		return nil, trace.BadParameter("no identity provider certificate provided, either set certificate as a parameter or via entity_descriptor")
+	}
+
+	if sc.GetSigningKeyPair() == nil {
+		keyPEM, certPEM, err := utils.GenerateSelfSignedSigningCert(pkix.Name{
+			Organization: []string{"Teleport OSS"},
+			CommonName:   "teleport.localhost.localdomain",
+		}, nil, 10*365*24*time.Hour)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		sc.SetSigningKeyPair(&SigningKeyPair{
+			PrivateKey: string(keyPEM),
+			Cert:       string(certPEM),
+		})
+	}
+
+	keyStore, err := utils.ParseSigningKeyStorePEM(sc.GetSigningKeyPair().PrivateKey, sc.GetSigningKeyPair().Cert)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	log.Debugf("[SAML] SSO: %v", sc.GetSSO())
+	log.Debugf("[SAML] Issuer: %v", sc.GetIssuer())
+	log.Debugf("[SAML] ACS: %v", sc.GetAssertionConsumerService())
+
+	sp := &saml2.SAMLServiceProvider{
+		IdentityProviderSSOURL:         sc.GetSSO(),
+		IdentityProviderIssuer:         sc.GetIssuer(),
+		ServiceProviderIssuer:          sc.GetServiceProviderIssuer(),
+		AssertionConsumerServiceURL:    sc.GetAssertionConsumerService(),
+		SignAuthnRequests:              true,
+		SignAuthnRequestsCanonicalizer: dsig.MakeC14N11Canonicalizer(),
+		AudienceURI:                    sc.GetAudience(),
+		IDPCertificateStore:            &certStore,
+		SPKeyStore:                     keyStore,
+		Clock:                          dsig.NewFakeClock(clock),
```

Comment:
Why is this wrapping the clock into a fake clock? This does not look like a test method - am I missing anything?

Reply:
The function `NewFakeClock` actually just wraps the clock as a `dsig.Clock`. So if `clock` is a real clock, this is equivalent to `dsig.NewRealClock()`.